src/core/vm.c

```diff
@@ -225,6 +228,55 @@ static void vm_init_dev(struct vm* vm, const struct vm_config* vm_config)
     }
 }
 
+static void vm_init_remio(struct vm* vm, const struct vm_config* vm_config)
+{
+    if (vm_config->platform.remio_dev_num == 0) {
+        return;
+    }
+
+    vm->remio_dev_num = vm_config->platform.remio_dev_num;
+    vm->remio_devs = vm_config->platform.remio_devs;
+
+    for (size_t i = 0; i < vm_config->platform.remio_dev_num; i++) {
+        struct remio_dev* remio_dev = &vm_config->platform.remio_devs[i];
+        struct shmem* shmem = shmem_get(remio_dev->shmem.shmem_id);
+        if (shmem == NULL) {
+            WARNING("Invalid shmem id in configuration. Ignored.");
+            continue;
+        }
+        size_t shmem_size = remio_dev->shmem.size;
+        if (shmem_size > shmem->size) {
+            shmem_size = shmem->size;
+            WARNING("Trying to map region to smaller shared memory. Truncated");
+        }
+        spin_lock(&shmem->lock);
+        shmem->cpu_masters |= (1UL << cpu()->id);
+        spin_unlock(&shmem->lock);
+
+        struct vm_mem_region reg = {
+            .base = remio_dev->shmem.base,
+            .size = shmem_size,
+            .place_phys = true,
+            .phys = shmem->phys,
+            .colors = shmem->colors,
+        };
+
+        vm_map_mem_region(vm, &reg);
+
+        if (remio_dev->type == REMIO_DEV_FRONTEND) {
+            struct emul_mem* emu = objpool_alloc(&emul_cache);
+            if (emu == NULL) {
+                ERROR("Failed allocating emulation memory node");
+            }
+            emu->va_base = remio_dev->va;
+            emu->size = remio_dev->size;
+            emu->handler = remio_mmio_emul_handler;
+            vm_emul_add_mem(vm, emu);
+        }
```

Comment:
move to a new function `vm_init_remio_dev`

Reply:
Update: https://github.com/bao-project/bao-hypervisor/pull/176/commits/7e8cd56617b25b56b88306f43ab9c491286c3764